src/librustc_span/lib.rs

```diff
@@ -81,11 +81,58 @@ impl Globals {
 
 scoped_tls::scoped_thread_local!(pub static GLOBALS: Globals);
 
+/// FIXME: Perhaps this should not implement Rustc{Decodable, Encodable}
+#[derive(Debug, Eq, PartialEq, Clone, Ord, PartialOrd, Hash, RustcDecodable, RustcEncodable)]
+#[derive(HashStable_Generic)]
+pub enum RealFileName {
+    Named(PathBuf),
+    /// For de-virtualized paths (namely paths into libstd that have been mapped
+    /// to the appropriate spot on the local host's file system),
+    Devirtualized {
+        /// `local_path` is the (host-dependent) local path to the file.
+        local_path: PathBuf,
+        /// `virtual_name` is the stable path rustc will store internally within
+        /// build artifacts.
+        virtual_name: PathBuf,
+    },
```

Comment:
I didn't consider this but I guess it makes sense. If we're willing to do this, we can just remove the `/rustc/.../` hack of encoding it as a path and just rely on this `enum`. (You can open an issue about this and/or leave a FIXME in the code if you want)

Reply:
Great! I was musing about such a change in my head, so I'm glad to hear you had similar thoughts. I'll add the FIXME; there's already so many others in this PR, another won't hurt.